migrations/files.js

```diff
@@ -0,0 +1,225 @@
+	},
+	refPermModel: {
+		type: String,
+		enum: ['user', 'role']
+	},
+	write: { type: Boolean, default: true },
+	read: { type: Boolean, default: true },
+	create: { type: Boolean, default: true },
+	delete: { type: Boolean, default: true },
+}, { _id: false});
+
+const fileSchema = new Schema({
+	isDirectory: { type: Boolean, default: false },
+	name: { type: String },
+	size: { type: Number },
+	type: { type: String },
+	storageFileName: { type: String },
+	thumbnail: { type: String },
+	shareToken: { type: String },
+	parent: { type: Schema.Types.ObjectId, ref: 'file' },
+	owner: {
+		type: Schema.Types.ObjectId,
+		required: true,
+		refPath: 'refOwnerModel'
+	},
+	refOwnerModel: {
+		type: String,
+		required: true,
+		enum: ['user', 'course', 'teams']
+	},
+	permissions: [permissionSchema],
+	lockId: { type: Schema.Types.ObjectId, ref: 'user' },
+	createdAt: { type: Date, 'default': Date.now },
+	updatedAt: { type: Date, 'default': Date.now }
+});
+
+const FileModel = mongoose.model('file', fileSchema, 'files');
+
+const run = async (dry) => {
+
+	const { _id: studentRoleId } = await RoleModel.findOne({ name: 'student' }).exec();
+
+	const logGreen = obj => {
+		if( dry ) {
+			console.log(chalk.green(obj));
+		}
+	};
+
+	const convertDocument = doc => {
+		logGreen(`Converting document ${doc.name} with path ${doc.path}`);
+
+		const [refOwnerModel, owner, ] = doc.key.split('/');
+
+		const refOwnerModelMap = {
+			'users': 'user',
+			'courses': 'course',
+		};
+
+		let permissions = [];
+
+		if( refOwnerModel === 'users' ) {
+			permissions.push({
+				refId: owner,
+				refPermModel: 'user',
+				write: true,
+				read: true,
+				create: true,
+				delete: true,
+			});
+		}
+
+		if( doc.permissions && doc.permissions.length ) {
+			permissions = [...permissions, ...doc.permissions.map(perm => ({
+				refId: perm.userId,
+				refPermModel: 'user',
+				write: perm.permissions.indexOf('can-write') > -1,
+				read: perm.permissions.indexOf('can-read') > -1,
+				create: false,
+				delete: false,
+			}))];
+		}
+
+		// Props obsolete in new model
+		const nullers = {
+			_id: undefined,
+			__v: undefined,
+			key: undefined,
+			path: undefined,
+			schoolId: undefined,
+			flatFileName: undefined
+		};
+
+		if( doc.studentCanEdit ) {
+			permissions.push({
+				refId: studentRoleId,
+				refPermModel: 'role',
+				write: Boolean(doc.studentCanEdit),
+				read: Boolean(doc.studentCanEdit),
+				create: false,
+				delete: false,
+			});
+		}
+
+		return sanitizeObj({
+			...doc,
+			isDirectory: !doc.type,
+			refOwnerModel: refOwnerModelMap[refOwnerModel] || refOwnerModel,
+			owner,
+			storageFileName: doc.flatFileName,
+			permissions,
+			...nullers,
+		});
+	};
+
+	const spawnDocuments = (directories, parent) => {
+		let transformed = directories.map(convertDocument);
+
+		if( parent ) {
+			transformed = transformed.map(doc => ({...doc, parent}));
+		}
+
+		const promises = transformed.map(d => dry ? d : FileModel.create(d));
```

Comment:
Is it okay to have all new file meta-objects in memory at once? I bet there are a lot of them.